[mono] Fix `LLVMArgWasmVtypeAsScalar` when a method doesn't set its return value

kotlarmilos opened the pull request:
## Description

This PR should fix an assertion in cases when a method doesn't set its return value. Additionally, it should fix trimming-related issues introduced in https://github.com/dotnet/runtime/issues/101104#issuecomment-2069238760. 

Fixes https://github.com/dotnet/runtime/issues/101104

kotlarmilos committed:
Fix an assertion when a method doesn't set its return value

kotlarmilos commented:
/azp run runtime-extra-platforms

kotlarmilos commented:
/azp run runtime-extra-platforms-wasm

kotlarmilos commented:
/azp run runtime-wasm

radekdoulik commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -6189,11 +6189,11 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 					break;
 				case LLVMArgWasmVtypeAsScalar:
-					if (!addresses [ins->sreg1]) {
+					if (!addresses [ins->sreg1] && lhs) {
```
Shouldn't this case, when the methods just throw, be handled here? https://github.com/dotnet/runtime/pull/101299/files#diff-cc4f4d3a653b6ac0bc4a6a001e4f0321e94ce51cc16ec85dc18675d1f8976365R6124-R6131

kotlarmilos replied:
You are right, but the condition seems te be false in case of `LLVMArgWasmVtypeAsScalar` due to `src_in_reg=FALSE` in:
```c
if (src_in_reg && (!lhs || ctx->is_dead [ins->sreg1])) {
```

Do you think that we should update `case LLVMArgNormal: src_in_reg = TRUE; break;` to include `LLVMArgWasmVtypeAsScalar`?

radekdoulik replied:
I think we should handle it similar to `LLVMArgVtypeAsScalar`, see my commit. (github didn't want me to allow to commit it as suggestion, so I committed it directly)

Then the rest of the changes should not be needed.

kotlarmilos replied:
Thanks for help!

kotlarmilos committed:
Test LLVMBuildRet ins for LLVMArgWasmVtypeAsScalar when lhs is null

kotlarmilos commented:
/azp run runtime-wasm

radekdoulik committed:
Update mini-llvm.c

I think we should handle it here, the same way as `LLVMArgVtypeAsScalar`. They differ only in the return type, where the wasm specific one is returning type of the element instead of `int`.

kotlarmilos committed:
Revert changes

kotlarmilos commented:
/azp run runtime-wasm

kotlarmilos committed:
Include TrimmerRootDescriptor only on Apple mobile platforms

kotlarmilos commented:
/azp run runtime-wasm

kotlarmilos commented:
/azp run runtime-extra-platforms

lambdageek commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -6119,7 +6119,7 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 				switch (linfo->ret.storage) {
 				case LLVMArgNormal: src_in_reg = TRUE; break;
-				case LLVMArgVtypeInReg: case LLVMArgVtypeAsScalar: src_in_reg = is_simd; break;
+				case LLVMArgVtypeInReg: case LLVMArgVtypeAsScalar: case LLVMArgWasmVtypeAsScalar: src_in_reg = is_simd; break;
```
```suggestion
				case LLVMArgVtypeInReg:
				case LLVMArgVtypeAsScalar:
				case LLVMArgWasmVtypeAsScalar:
					src_in_reg = is_simd;
					break;
```

follow-up idea: this function with a ton of nested switch statements is hard to think about.  Let's make helper methods.

kotlarmilos replied:
https://github.com/dotnet/runtime/issues/101412

kotlarmilos committed:
Update src/mono/mono/mini/mini-llvm.c

Co-authored-by: Aleksey Kliger (λgeek) <akliger@gmail.com>

kotlarmilos committed:
Fix formatting in mini-llvm.c

kotlarmilos commented:
/azp run runtime-wasm

kotlarmilos commented:
/azp run runtime-extra-platforms

kotlarmilos committed:
Update System.Reflection.MetadataLoadContext.Tests.csproj to conditionally include TrimmerRootDescriptor on Apple mobile platforms

kotlarmilos commented:
/azp run runtime-wasm,runtime-extra-platforms

kotlarmilos commented:
According to the build analysis, the failures are known and shouldn't be related. The wasm-extra-platforms failures seem to be the same as in https://dev.azure.com/dnceng-public/public/_build?definitionId=151&_a=summary.

I've created a tracking issue: https://github.com/dotnet/runtime/issues/101482
